pkg/server/plugin/datastore/sql/sql.go

```diff
@@ -1124,149 +1119,861 @@ func createRegistrationEntry(tx *gorm.DB,
+
+	isSubset := func(ss []*common.Selector) bool {
+		for _, s := range ss {
+			if _, ok := set[selectorKey{Type: s.Type, Value: s.Value}]; !ok {
+				return false
 			}
-		case datastore.BySelectors_MATCH_EXACT:
-			selectorsList = append(selectorsList, selectorSet.Raw())
-		default:
-			return nil, fmt.Errorf("unhandled match behavior %q", req.BySelectors.Match)
 		}
+		return true
 	}
 
-	// filter registration entries
-	entryTx := tx
-	if req.ByParentId != nil {
-		entryTx = entryTx.Where("parent_id = ?", req.ByParentId.Value)
+	filtered := make([]*common.RegistrationEntry, 0, len(entries))
+	for _, entry := range entries {
+		if isSubset(entry.Selectors) {
+			filtered = append(filtered, entry)
+		}
 	}
-	if req.BySpiffeId != nil {
-		entryTx = entryTx.Where("spiffe_id = ?", req.BySpiffeId.Value)
+	return filtered
+}
+
+func listRegistrationEntriesOnce(ctx context.Context, dbType string, db *sql.DB, req *datastore.ListRegistrationEntriesRequest) (*datastore.ListRegistrationEntriesResponse, error) {
+	query, args, err := buildListRegistrationEntriesQuery(dbType, req)
+	if err != nil {
+		return nil, sqlError.Wrap(err)
 	}
 
-	if len(selectorsList) == 0 {
-		// no selectors to filter against.
-		var entries []RegisteredEntry
-		entries, p, err = findRegisteredEntries(entryTx, req.Pagination)
-		if err != nil {
-			return nil, sqlError.Wrap(err)
+	rows, err := db.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, sqlError.Wrap(err)
+	}
+	defer rows.Close()
+
+	var entries []*common.RegistrationEntry
+	if req.Pagination != nil {
+		entries = make([]*common.RegistrationEntry, 0, req.Pagination.PageSize)
+	} else {
+		// start the slice off with a little capacity to avoid the first few
+		// reallocations
+		entries = make([]*common.RegistrationEntry, 0, 64)
+	}
+	var lastEID uint64
+
+	for rows.Next() {
+		var r entryRow
+		if err := scanEntryRow(rows, &r); err != nil {
+			return nil, err
 		}
 
-		respEntries, err := modelsToEntries(tx, entries)
-		if err != nil {
-			return nil, sqlError.Wrap(err)
+		var entry *common.RegistrationEntry
+		if len(entries) == 0 || lastEID != r.EId {
+			lastEID = r.EId
+			entry = new(common.RegistrationEntry)
+			entries = append(entries, entry)
+		} else {
+			entry = entries[len(entries)-1]
 		}
 
-		return &datastore.ListRegistrationEntriesResponse{
-			Entries:    respEntries,
-			Pagination: p,
-		}, nil
+		if err := fillEntryFromRow(entry, &r); err != nil {
+			return nil, err
+		}
 	}
 
-	modelsSet := make(map[uint]RegisteredEntry)
-	for _, selectors := range selectorsList {
-		refCount := make(map[uint]int)
-		for _, s := range selectors {
-			var results []Selector
-			if err := tx.Find(&results, "type = ? AND value = ?", s.Type, s.Value).Error; err != nil {
-				return nil, sqlError.Wrap(err)
+	if err := rows.Err(); err != nil {
+		return nil, sqlError.Wrap(err)
+	}
+
+	resp := &datastore.ListRegistrationEntriesResponse{
+		Entries: make([]*common.RegistrationEntry, 0, len(entries)),
```

Comment:
why do we need to make a new list?

Reply:
Oh, this is leftover from when we were aggregating into a map. Will remove.